### ISSUE TITLE
Proposal for a new v5 options scheme to allow for all kinds of flexibility

### ISSUE THREAD

**Issue by davidtheclark** [davidtheclark/cosmiconfig#126]
We've had quite a few requests, some recent and some older, asking for more flexible ways to find, name, and parse files. Here's what I dug up:

- [ES2015 modules](https://github.com/davidtheclark/cosmiconfig/issues/109)
- [Whatever ES2015+ syntax Babel can handle](https://github.com/davidtheclark/cosmiconfig/issues/109)
- [Alternate JS extensions](https://github.com/davidtheclark/cosmiconfig/issues/122)
- [TypeScript](https://github.com/davidtheclark/cosmiconfig/issues/122)
- [JSON5](https://github.com/davidtheclark/cosmiconfig/issues/119)
- [`.config.js` files starting a dot](https://github.com/davidtheclark/cosmiconfig/issues/57), like `.foo.config.js`
- [A subproperty in `package.json` nested under a `config` property](https://github.com/davidtheclark/cosmiconfig/issues/56), like `{"config":{"foo":{..}}}`
- Searching [within a `.config/` subdirectory of each searched directory](https://github.com/davidtheclark/cosmiconfig/issues/50)

We've also had [a discussion](https://github.com/davidtheclark/cosmiconfig/pull/124) about the way extensionless rc files are parsed. We've talked about adding a `parseJs` option (to address the request above for alternate flavors of JS); and, as was inevitable, the suggestion came up for `parseJson` and `parseYaml` options to allow for alternate flavors of *those* languages — which got me anticipating eventual requests to parse XML or INI or yet-unheard-of languages. It also got me looking at the current `rc`, `rcStrictJson`, and `rcExtensions` options, and realizing how conceptually tangled they are.

So, **I'm wondering if there's a way we could set up v5 to allow for *any filenames*, searched for in *any order*, and parsed with *any parsers*.**

Here's what I've come up with:

What if we do away with most of the existing options (`packageProp`, `rc`, `js`, `rcStrictJson`, and `rcExtensions`), and instead have a flexible `searchScheme` option. 

- `searchScheme` is an array representing an ordered list of things to search for in each directory.
- Each item in the array is either an object or a string.
- Object items can have the following properties:
  - `filename`: The filename of a file to search for, relative to the directory being searched; e.g. `.stylelintrc`, `.stylelintrc.json`, `stylelint.config.js`, `package.json`. Could even include a directory if you wanted to search in a `.config/` subdirectory of every searched directory.
  - `parser`: A parser function to use. *Optional:* If this is not provided, we'll parse `.js` as JS, `.json` as JSON, `.yaml` and `.yml` as YAML, and extensionless rc files as YAML (which incorporates JSON). We'd expose some standard parsers on the `cosmiconfig` object that people could use here, or they could bring their own.
  - `property`: A property to look for in the file. *Optional:* The main use case for this is `package.json`, and by default it will match the `moduleName` you provided when you initialized cosmiconfig. We could even accept an array to indicate nesting, to address the request above for `config.mything` within `package.json`.
- String items are the equivalent of providing only a `filename` on an object and accepting the other defaults.

Here are a bunch of examples to illustrate how `searchScheme` would work to address all these use cases. They all assume your tool's name is `plod`.

```js
// The current default settings, verbose.
a = [
  {
    filename: 'package.json',
    property: 'plod',
    parser: cosmiconfig.parseJson
  },
  {
    filename: '.plodrc',
    parser: cosmiconfig.parseYaml
  },
  {
    filename: 'plod.config.js',
    parser: cosmiconfig.parseJs
  }
]

// The current default settings, concise.
b = [
  'package.json',
  '.plodrc',
  'plog.config.js'
]

// The current default plus rcStrictJson: true,
// so not allowing YAML in extensionless rc files.
c = [
  'package.json',
  {
    filename: '.plodrc',
    parser: cosmiconfig.parseJson
  },
  'plod.config.js'
]

// The current default plus rcExtensions: true,
// with string entries. **I suggest we make this
// the new default.**
d = [
  'package.json',
  '.plodrc',
  '.plodrc.json',
  '.plodrc.yaml',
  '.plodrc.yml',
  '.plodrc.js',
  'plod.config.js'
]

// ESLint could do this.
// cf. https://eslint.org/docs/user-guide/configuring#configuration-file-formats
e = [
  '.eslintrc.js',
  '.eslintrc.yaml',
  '.eslintrc.yml',
  '.eslintrc.json',
  '.eslintrc',
  {
    filename: 'package.json',
    property: 'eslintConfig'
  }
]

// Babel could do this.
// cf. https://babeljs.io/docs/usage/babelrc/
f = [
  'package.json',
  {
    filename: '.babelrc',
    parser: someJson5Parser
  }
]

// Allow a JS file with ES2015 module syntax
// and .js or .mjs extensions; or a .ts file;
g = [
  {
    filename: 'plod.js',
    parser: someEs2015ModuleParser
  },
  {
    filename: 'plod.mjs',
    parser: someEs2015ModuleParser
  },
  {
    filename: 'plod.ts',
    parser: someTypeScriptParser
  }
]

// Allow extensionless YAML and JSON files
// within or outside of .config/ subdirectories.
h = [
  {
    filename: '.plodrc',
    parser: cosmiconfig.parseYaml
  },
  {
    filename: '.config/.plodrc',
    parser: cosmiconfig.parseYaml
  }
]

// Look within the config property of package.json.
i = [
  {
    filename: 'package.json',
    property: ['config', 'plod']
  }
]
```

This proposal wouldn't affect the way `load`/`loadSync` and `search`/`searchSync` or the caches work — just the way you configure cosmiconfig to search for and parse the formats you want. I think it could work, and also provide the possibility for indefinite extension without the need to add more options in the future. (At most, we might end up adding *more defaults* or exposing *more parsers* in the future, if we find that some unaddressed usage is super common.)

I suggest that we add rc files with extensions to the default configuration because I think it's being used all over the place. I know that stylelint, Prettier, and lint-staged use it.

**Comment by olsonpm** [comment]
i like this, and agree with the rc+extension default.  I'm only unsure whether it addresses the alternative directories like `.config` because I assume users will expect the ability to search for this list of files under alternative directories.  If i'm understanding your new API, users would have to declare all the files individually underneath custom directories.  This would be merely an inconvenience though, and I can't think of a solution that doesn't add overly complex configurability.

**Comment by davidtheclark** [comment]
> If i'm understanding your new API, users would have to declare all the files individually underneath custom directories.

Yep. I agree that this isn't much of a problem, though.

Here's another example I forgot to add:

```js
// Some dark future ...
j = [
  {
    filename: '.plodconfig',
    parser: someIniParser
  },
  {
    filename: 'plod_config.xml',
    parser: someXmlParser
  },
  {
    filename: '.plodrc',
    parser: someUnheardOfLanguageParser
  },
  {
    filename: 'plod-config.csv',
    parser: someCsvParser
  }
]
```

**Comment by davidtheclark** [comment]
I realize that I hadn't thought much about [the desire to be able to use require hooks](https://github.com/davidtheclark/cosmiconfig/issues/109).

Each `parser` function would receive the file's path and code (string) as arguments. So I believe that our default JS "parser" would ignore the code and just `return require(filePath)`. An alternate parser could do the same kind of thing thing (e.g. `return require(''esm")(filePath)`); or it could use the code string (e.g. pass it through Babel).

That makes the word "parser" a little unsuitable, though, because the "parser" function might not *just* parse, but also load. We could use `loader` instead of `parser` in order to address that. Might be a little strange for all the use cases where you're *just parsing* the string that you receive; but it might also be a generic enough word that it's fine.

**Comment by davidtheclark** [comment]
I started implementing this idea in https://github.com/davidtheclark/cosmiconfig/pull/128 and realized that it will need to affect the `load` function, as well.

Maybe we can name the option `places` (instead of `searchSchema`), because it will apply to both `search` and `load`.

When using `load`, we'll check whether the file you're loading corresponds to any entry in `places`. If so, we'll use the loader determined by that entry (which might be custom). If not, we'll throw an error saying you're not allowed to load that file.

**Comment by olsonpm** [comment]
in v4, any file could be loaded and all the parsers would be attempted correct?  I'm wondering whether strictly checking `places` would conflict with user-defined config paths such as prettier's [`--config`](https://prettier.io/docs/en/cli.html#find-config-path-and-config)

> You can also use --config if your configuration file lives somewhere where prettier cannot find it, such as a config/ directory.

**Comment by davidtheclark** [comment]
@olsonpm: That is a very good and devastating point. This whole idea doesn't address `load` very well, does it. 🤔 

Maybe we should consider the following variation:

- Three options: `loaders`, `searchPlaces`, and `packageProp`.
- `loaders` maps extensions to loaders. `js`, `json`, `yaml`, `yml`, and `no-extension` all have defaults. Users could set `js` to their Babel-powered loader, or `mjs` to their ES2015 module loader, or `ts` to their TypeScript loader, or `ini` to their INI loader. They could also set `no-extension` to `cosmiconfig.loadJson` if they want to only allow strict JSON. `loaders` can apply to both `search` and `load`.
- `searchPlaces` is just an array of paths (no objects), determining which files we look at, in which order. It only applies to `search`.
- `packageProp` determines which property to look for in `package.json`, same as now. We won't build in the ability to get nested properties on other config files, but users could do that with custom loaders.

I think this plan would address all the use cases described above and also handle `load`.

**Comment by olsonpm** [comment]
makes sense to me.  The only possible issue I can see is if libraries allow for `.ini` configurations (which commonly don't have extensions), then not being able to check against multiple loaders for `no-extension`.  However I still prefer keeping the api simple and not allowing an array of loaderes for `no-extension` until demand calls for it.

**Comment by davidtheclark** [comment]
Implemented in #129.
